Add --grab-reboot option to trigger system reboot via keyboard shortcut

jtollet opened the pull request:
## Summary

This PR adds a new `--grab-reboot` option that allows users to configure a keyboard shortcut to reboot the system directly from kmscon.

## Features

- **Disabled by default**: Users must explicitly configure a key combination
- **Direct system call**: Uses `reboot(RB_AUTOBOOT)` for reliability
- **Configurable**: Supports any key combination (e.g., `<Ctrl><Alt>Delete`)

## Implementation

- Added `grab_reboot` field to `kmscon_conf_t` configuration structure
- Implemented `seat_trigger_reboot()` function using `reboot(RB_AUTOBOOT)`
- Integrated into seat-level input event handling in `seat_input_event()`
- Updated `kmscon.1.xml.in` and `kmscon.conf.1.xml.in` man pages

## Usage

```bash
# Command line
kmscon --grab-reboot='<Ctrl><Alt>Delete'

# Configuration file
grab-reboot=<Ctrl><Alt>Delete
```

## Modified Files

- `src/kmscon_conf.h` - Configuration structure
- `src/kmscon_conf.c` - Option definition and help text
- `src/kmscon_seat.c` - Reboot trigger implementation
- `docs/man/kmscon.1.xml.in` - Command-line option documentation
- `docs/man/kmscon.conf.1.xml.in` - Configuration file documentation

michael-oberpriller commented on src/kmscon_seat.c:
```diff
@@ -32,6 +32,9 @@
 #include <stdlib.h>
 #include <string.h>
+#include <unistd.h>
+#include <sys/wait.h>
```
Pretty sure <sys/wait.h> is not necessary.

jtollet replied:
Good point. I fixed it.

michael-oberpriller commented on src/kmscon_seat.c:
```diff
@@ -634,7 +649,13 @@ static void seat_input_event(struct uterm_input *input, struct uterm_input_key_e
+		seat_trigger_reboot(seat);
+		return;
+	}
+	}
```
Please fix whitespace.

jtollet commented:
I confirm this PR is good.

De : Jerome Tollet (jtollet) ***@***.***>
Date : dimanche, 14 décembre 2025 à 23:18
À : kmscon/kmscon ***@***.***>, kmscon/kmscon ***@***.***>
Cc : Author ***@***.***>
Objet : Re: [kmscon/kmscon] Add --grab-reboot option to trigger system reboot via keyboard shortcut (PR #176)

Wait a second before merging. I am investigating one more thing.

De : Jocelyn Falempe ***@***.***>
Date : dimanche, 14 décembre 2025 à 22:57
À : kmscon/kmscon ***@***.***>
Cc : Jerome Tollet (jtollet) ***@***.***>, Author ***@***.***>
Objet : Re: [kmscon/kmscon] Add --grab-reboot option to trigger system reboot via keyboard shortcut (PR #176)

[https://avatars.githubusercontent.com/u/554627?s=20&v=4]kdj0c left a comment (kmscon/kmscon#176)<https://github.com/kmscon/kmscon/pull/176#issuecomment-3652244403>

I'm a bit skeptical about this feature, but the code looks good, and it doesn't introduce complexity, so if no other comments I will merge it soon.

—
Reply to this email directly, view it on GitHub<https://github.com/kmscon/kmscon/pull/176#issuecomment-3652244403>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGBK27R7CIJ2AF3FWB25ZM34BXMK3AVCNFSM6AAAAACO6JNWRSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMNJSGI2DINBQGM>.
You are receiving this because you authored the thread.

jtollet commented:
@satmandu that quite common... And that's why linux vt supports it too. TY !

Karlson2k commented:
While the feature is very useful, this implementation is quite dangerous.
With this implementation, a syscall (via a libc wrapper) is called to perform the reboot directly.
In contrast, when Ctrl-Alt-Del is pressed in the Linux console, `SIGINT` is sent to PID 1, and then the init process performs a clean reboot.
I think people who are using this feature are expecting the same behaviour as with the standard Linux console, and they can be surprised by data loss or a damaged filesystem due to an unclean shutdown.

I suggest either replacing this hard reboot with a proper `SIGINT` to PID 1, or at least making this configurable.
In the meantime, a **BIG FAT** warning could be added to the configuration example to avoid unpleasant surprises for users.

jtollet commented:
@Karlson2k Thank you for the excellent feedback!

You're absolutely right about the concerns with hard reboot. I've addressed all your suggestions in PR #260:

**What was implemented:**
- **SIGINT to PID 1 (soft mode)** is now the **default behavior**, matching the standard Linux console Ctrl-Alt-Del
- **Hard mode is optional** via `--grab-reboot-mode=hard`
- **BIG FAT warnings** added to all documentation (man pages and config example) about data loss and filesystem corruption risks with hard mode
- Clear documentation that soft mode matches what users expect from the Linux console

**Testing:**
Both modes have been tested on real hardware:
- ✅ Soft mode: clean shutdown via SIGINT to PID 1 (default)
- ✅ Hard mode: sync() + immediate reboot via RB_AUTOBOOT (optional, with warnings)

The implementation follows your recommendation to make this configurable while defaulting to the safe behavior users expect.

See PR #260 for the documentation updates: https://github.com/kmscon/kmscon/pull/260
